### ISSUE TITLE
watchSource regression: adding watched files in subdirs do not trigger ~ tasks anymore

### ISSUE THREAD

**Issue by bjornregnell** [sbt/sbt#4272]
Adding file extensions to watch using `watchSources ++= ((baseDirectory.value / "some-dir") * "*.txt").get` works in sbt version 0.13.x but not in 1.1.6.

## steps
```
$ ls
build.sbt  project  some-dir  target

$ cat some-dir/some-file.txt
hello

$ cat build.sbt
lazy val hello = taskKey[Unit]("print hello")
hello := println("Hello triggered!")
watchSources ++= ((baseDirectory.value / "some-dir") * "*.txt").get

$ sbt
[info] Loading project definition from /home/bjornr/tmp/sbt-test/project
[info] Set current project to sbt-test (in build file:/home/bjornr/tmp/sbt-test/)
> sbtVersion
[info] 0.13.17
> ~hello
Hello triggered!
[success] Total time: 0 s, completed Jul 17, 2018 1:43:14 PM
1. Waiting for source changes... (press enter to interrupt)
Hello triggered!
[success] Total time: 0 s, completed Jul 17, 2018 1:43:31 PM
2. Waiting for source changes... (press enter to interrupt)
Hello triggered!
[success] Total time: 0 s, completed Jul 17, 2018 1:43:34 PM
3. Waiting for source changes... (press enter to interrupt)
Hello triggered!
[success] Total time: 0 s, completed Jul 17, 2018 1:43:37 PM
4. Waiting for source changes... (press enter to interrupt)

> reboot
[info] Loading project definition from /home/bjornr/tmp/sbt-test/project
[info] Loading settings from build.sbt ...
[info] Set current project to sbt-test (in build file:/home/bjornr/tmp/sbt-test/)
[info] sbt server started at local:///home/bjornr/.sbt/1.0/server/cc4007fc9be5d5a34053/sock
sbt:sbt-test> sbtVersion
[info] 1.1.6
sbt:sbt-test> ~hello
Hello triggered!
[success] Total time: 0 s, completed Jul 17, 2018 1:44:20 PM
1. Waiting for source changes... (press enter to interrupt)

// AND NOW IN 1.1.6 THE TASK IS NOT TRIGGERED
// although the file is in there:

sbt:sbt-test> show watchSources
...
...
[info] * Source(
[info]   base = /home/bjornr/tmp/sbt-test/some-dir/some-file.txt,
[info]   includeFilter = sbt.io.AllPassFilter$@591f6f83,
[info]   excludeFilter = sbt.io.NothingFilter$@2b44605c,
[info]   recursive = true,
[info] )

$ cat /proc/version
Linux version 4.15.0-24-generic (buildd@lcy01-amd64-010) (gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.9)) #26~16.04.1-Ubuntu SMP Fri Jun 15 14:35:08 UTC 2018
```

## problem

Custom task is not triggered by file update on watchSources with added file extensions in subdirs and this works in 0.13.x but not in 1.1.6. 

## expectation

The custom `~hello` task above should be triggered by file update in 1.1.6 just as in 0.13.17.

## notes

It works if the file is put on project dir top level and added as `watchSources ++= (baseDirectory.value * "*.txt").get` but it does not work if put in a subdir as in the steps above.

sbt version: *1.1.6*

**Comment by eed3si9n** [comment]
Thanks for the report with repro steps!

**Comment by eatkins** [comment]
That is a bug, but a workaround is to change your build.sbt to: `watchSources += WatchSource(baseDirectory.value / "some-dir", "*.txt", NothingFilter)`

**Comment by eatkins** [comment]
Also, the `.get` syntax for specifying watch sources should be avoided moving forward because it causes unnecessary file system access compared to the workaround I just posted.

**Comment by bjornregnell** [comment]
Many thanks @eatkins for help with workaround and info on avoiding get! Is this workaround perhaps mentioned somewhere (where?) in the docs? Couldn't find it but I might be looking in the wrong place.

**Comment by bjornregnell** [comment]
The workaround by @eatkins worked like a charm both in 1.1.6 and 1.2.0. 
I made this little helper to cut down on boilerplate as I have several dirs and many different file filters to include:
```scala
def ws(base: sbt.io.syntax.File, include: sbt.io.FileFilter): sbt.internal.io.Source =
  WatchSource(base, include, excludeFilter=NothingFilter)

watchSources += ws(baseDirectory.value / "some-dir",    "*.tex")
watchSources += ws(baseDirectory.value / "another-dir", "*.cls")
// etc.
```
I'm now happily riding the 1.2.0 train and sbt is faster and smoother than ever :+1: 
@eed3si9n Thanks to the sbt team for a great new release!

**Comment by eatkins** [comment]
@bjornregnell glad to hear that things are working for you. I didn't add any documentation personally for the new WatchSource approach. I regret that the excludeFilter parameter exists because I don't really see the purpose of it (since you can just `&&` the include filter with the excludes). I've been swamped on a different project recently, but that is very close to completion, so hopefully I'll have time to take a look at the documentation soonish.

**Comment by bjornregnell** [comment]
> I regret that the excludeFilter parameter exists

@eatkins Perhaps another `apply` on `WatchSources` w/o the param or a default value `excludeFilter=NothingFilter` in the existing `apply` could help cut down on the boilerplate w/o breaking existing code?

> I'll have time to take a look at the documentation soonish

That'd be great! I find the example-based documentation with short code answering "how-to" questions the easiest way for me to get things done with sbt. Sometimes I find stuff by searching the complete pdf with ctrlf+f as well, but that's when I have a clue of what things are called... I also try to look at each release notes as those often include useful snippets.

**Comment by eatkins** [comment]
I believe this has been fixed in 1.3.x.
